presentation-api/controlling-ua/getAvailability.html

```diff
@@ -1,69 +1,53 @@
-                .then(function (availability) {
-                    assert_true(availability instanceof PresentationAvailability);
-                });
-    }, "The promise is an instance of PresentationAvailability");
-
-    // Availability.value is set Test
-    promise_test(function () {
-        var request = createRequestObject();
+<script>
 
-        return request.getAvailability()
-                .then(function (availability) {
-                    assert_true(typeof availability.value == 'boolean');
-                });
-    }, "The availability has an boolean value.");
+    // ---------------------------------------
+    // Presentation Availability Tests - begin
+    // ---------------------------------------
 
+    promise_test(function(t) {
+        var availability;
 
-    // Best Case Scenario Test
-    // -----------------------
-    promise_test(function () {
-        var request = createRequestObject();
+        // instance of PresentationRequest
+        var request = new PresentationRequest(presentationUrls);
+        assert_true(request instanceof PresentationRequest, 'The request is an instance of PresentationRequest.');
+
+        var promise = request.getAvailability();
+        // if the PresentationRequest object has an unsettled Promise, getAvailability returns that Promise
+        assert_equals(promise, request.getAvailability(), 'If the PresentationRequest object has an unsettled Promise, getAvailability returns that Promise.');
+
+        return promise.then(function (a) {
+            availability = a;
+
+            // instance of PresentationAvailability
+            assert_true(availability instanceof PresentationAvailability, 'The promise is resolved with an instance of PresentationAvailability.');
+            // PresentationAvailability.value is set as boolean
+            assert_equals(typeof availability.value, 'boolean', 'The availability has an boolean value.');
+            // PresentationAvailability.value is set to true when any presentation display is available
+            assert_true(availability.value, 'The availability value is true when any presentation display is available.');
+            // if the Promise of a previous call to getAvailability has already been settled, getAvailability returns a new Promise
+            var newPromise = request.getAvailability();
+            assert_not_equals(promise, newPromise, 'If the Promise from a previous call to getAvailability has already been settled, getAvailability returns a new Promise.');
+
+            return newPromise;
+        }, function(err) {
+            // getAvailability rejects a new Promise with NotSupportedError if the browser can find displays only when starting a connection
+            assert_equals(err.name, 'NotSupportedError', 'getAvailability() rejects a Promise with a NotSupportedError exception, if the browser can find presentation displays only when starting a connection.')
+        }).then(function (newAvailability) {
```

Comment:
I may be lost in promises, but given you catch a possible `getAvailability` failure before this `then`, isn't that "then" part also going to be run if a failure occurs? Catching the `NotSupportedError` error should end the test.

I'm not sure what is the cleanest way to write this:

- You could catch `NotSupportedError` in a final `.catch(function err) {}` block. That would look great, but would make the test succeed in the (unlikely) event when only the second call to getAvailability fails with a `NotSupportedError` error. I'd be fine with that personally, the implementation would be seriously buggy if it only rejects the second call.
- You could move that `.then` part to `return newPromise.then(function (newAvailability) { ... };`